File: src/json/caps_tenkey_mode.json.rb
require 'json'
require_relative '../lib/karabiner.rb'

TENKEY_MAPPING = {
  'm'=> '1',
  'comma'=> '2',
  'period'=> '3',
  'j'=> '4',
  'k'=> '5',
  'l'=> '6',
  'u'=> '7',
  'i'=> '8',
  'o'=> '9',
  'spacebar'=> '0'
}

ARITHMETIC_SYMBOLS_MAPPING = {
  'semicolon' => 'keypad_plus',
  'p' => 'keypad_hyphen',
  'slash'=> 'keypad_equal_sign',
  '9' => 'keypad_slash',
  '0' => 'keypad_asterisk'
}

def main
  puts JSON.pretty_generate(
    title: 'CAPS 4 Tenkey',
    maintainers: %w[IvanShamatov],
    rules: [
      tenkey_toggle,
      tenkey_mapping,
      arithmetic_symbols_mapping,
    ]
  )
end

def tenkey_toggle
  {
    description: 'CAPS Tenkey: TenkeyMode on/off toggle',
    manipulators: [
      # Turning on Tenkey mode
      {
        from: _from('caps_lock', [], ['any']),
        type: 'basic',
        to: _set_variable('tenkey_mode', 1),
        conditions: tenkey_mode_off
      },
      # Turning off Tenkey mode
      {
        from: _from('caps_lock', [], ['any']),
        type: 'basic',
        to: _set_variable('tenkey_mode', 0),
        conditions: tenkey_mode_on
      }
    ]
  }
end

def tenkey_mapping
  {
    description: 'CAPS Tenkey: if TenkeyMode ON m,.jkluio' ' maps to 1234567890',
    manipulators: map_keys(TENKEY_MAPPING, conditions: tenkey_mode_on)
  }
end

def arithmetic_symbols_mapping
  {
    description: 'CAPS Tenkey: if TenkeyMode ON 90p;/ maps to /*-+=',
    manipulators: map_keys(ARITHMETIC_SYMBOLS_MAPPING, conditions: tenkey_mode_on)
  }
end


def _from(key_code, mandatory = [], optional = [])
  {
    key_code: key_code,
    modifiers: {
      mandatory: mandatory,
      optional: optional
    }
  }
end

def map_keys(mapping, conditions: [])
  mapping.map do |from_key, to_key|
    {
      type: 'basic',
      from: _from(from_key),
      to: { key_code: to_key },
      conditions: conditions
    }
  end
end

def _set_variable(name, value)
  { set_variable: { name: name, value: value } }
end

def tenkey_mode_on
  [{ type: 'variable_if', name: 'tenkey_mode', value: 1 }]
end

def tenkey_mode_off
  [{ type: 'variable_if', name: 'tenkey_mode', value: 0 }]
end

main

File: src/json/caps_tenkey_mode.json.js
// JavaScript should be written in ECMAScript 5.1.

function main() {
  console.log(
    JSON.stringify(
      {
        title: 'CAPS 4 Tenkey',
        maintainers: ['IvanShamatov'],
        rules: [].concat(tenkeyToggle(), tenkeyMapping(), arithmeticSymbolsMapping()),
      },
      null,
      '  '
    )
  )
}

function tenkeyToggle() {
  return {
    description: 'CAPS Tenkey: TenkeyMode on/off toggle',
    manipulators: [
      // Turning on Tenkey mode
      {
        from: {
          key_code: 'caps_lock',
          modifiers: {
            optional: ['any'],
          },
        },
        type: 'basic',
        to: {
          set_variable: {
            name: 'tenkey_mode',
            value: 1,
          },
        },
        conditions: tenkeyModeOff(),
      },
      // Turning off Tenkey mode
      {
        from: {
          key_code: 'caps_lock',
          modifiers: {
            mandatory: [],
            optional: ['any'],
          },
        },
        type: 'basic',
        to: {
          set_variable: {
            name: 'tenkey_mode',
            value: 0,
          },
        },
        conditions: tenkeyModeOn(),
      },
    ],
  }
}

function tenkeyMapping() {
  const mapping = [
    { from: 'm', to: 'keypad_1' },
    { from: 'comma', to: 'keypad_2' },
    { from: 'period', to: 'keypad_3' },
    { from: 'j', to: 'keypad_4' },
    { from: 'k', to: 'keypad_5' },
    { from: 'l', to: 'keypad_6' },
    { from: 'u', to: 'keypad_7' },
    { from: 'i', to: 'keypad_8' },
    { from: 'o', to: 'keypad_9' },
    { from: 'spacebar', to: 'keypad_0' },
  ]

  return {
    description: 'CAPS Tenkey: if TenkeyMode ON m,.jkluio maps to keypad_[1234567890]',
    manipulators: mapKeys(mapping, tenkeyModeOn()),
  }
}

function arithmeticSymbolsMapping() {
  const mapping = [
    { from: 'semicolon', to: 'keypad_plus' },
    { from: 'p', to: 'keypad_hyphen' },
    { from: 'slash', to: 'keypad_equal_sign' },
    { from: '9', to: 'keypad_slash' },
    { from: '0', to: 'keypad_asterisk' },
  ]

  return {
    description: 'CAPS Tenkey: if TenkeyMode ON 90p;/ maps to /*-+=',
    manipulators: mapKeys(mapping, tenkeyModeOn()),
  }
}

function mapKeys(mapping, conditions) {
  return mapping.map(function (m) {
    return {
      type: 'basic',
      from: { key_code: m.from },
      to: { key_code: m.to },
      conditions: conditions,
    }
  })
}

function tenkeyModeOn() {
  return [{ type: 'variable_if', name: 'tenkey_mode', value: 1 }]
}

function tenkeyModeOff() {
  return [{ type: 'variable_if', name: 'tenkey_mode', value: 0 }]
}

main()
